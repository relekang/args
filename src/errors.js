// @flow
type CliErrorOptions = {
  message: string,
  exitCode: number,
  showHelp?: boolean
};
export class CliError extends Error {
  exitCode: number;
  showHelp: boolean;

  constructor({ message, exitCode, showHelp }: CliErrorOptions) {
    super(message);
    this.exitCode = exitCode;
    this.showHelp = showHelp || false;
  }

  toString() {
    return this.message;
  }
}
