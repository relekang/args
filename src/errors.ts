type CliErrorOptions = {
  message: string;
  exitCode: number;
  showHelp?: boolean;
};
export class CliError extends Error {
  exitCode: number;
  showHelp: boolean;

  constructor({ message, exitCode, showHelp }: CliErrorOptions) {
    super(message);
    this.exitCode = exitCode;
    this.showHelp = showHelp || false;

    Object.setPrototypeOf(this, CliError.prototype);
  }

  toString() {
    return this.message;
  }
}
