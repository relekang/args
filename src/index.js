// @flow

export type Config = {};

export default function args(_config: Config) {
  return ([_node, _program, ..._rest]: Array<String>) => ({});
}
