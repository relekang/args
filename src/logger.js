// @flow
/* eslint-disable no-console */

export function error(...args: *) {
  console.error(...args);
}

export function log(...args: *) {
  console.log(...args);
}
