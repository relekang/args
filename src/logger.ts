/* eslint-disable no-console, typescript/no-explicit-any */

export function error(...args: any[]) {
  console.error(...args);
}

export function log(...args: any[]) {
  console.log(...args);
}
