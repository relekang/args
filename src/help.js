// @flow
/* eslint-disable global-require */
import chalk from "chalk";
import wrap from "word-wrap";
import path from "path";
import fs from "fs";
import { promisify } from "util";

import type { Config, CommandConfig, CommandOption, Options } from "./types";

import * as logger from "./logger";

const readdir = promisify(fs.readdir);

function values<A>(o: { [key: string]: A }): Array<A> {
  // $FlowFixMe
  return Object.values(o);
}

async function getCommands(config): { [key: string]: CommandConfig } {
  if (config.commands) {
    return config.commands;
  }
  return (await readdir(config.commandsPath)).filter(/\.js$/.test).reduce(
    (lastValue, file) => ({
      ...lastValue,
      // $FlowFixMe
      [file.replace(/.js$/, "")]: require(path.join(config.commandsPath, file))
    }),
    {}
  );
}

export async function createCommandsList(config: Config) {
  const commands = await getCommands(config);

  const length = Math.max(
    0,
    ...values(commands).map(({ name }) => name.length)
  );
  return values(commands)
    .map(command => {
      const spacing = new Array(Math.max(1, length - command.name.length + 1))
        .fill(" ")
        .join("");
      return `${command.name}${spacing} - ${command.help || ""}`;
    })
    .join("\n");
}

export function createPositionalUsage(command: CommandConfig) {
  return (command.positionalOptions || [])
    .map(({ name, required }) => (required ? `<${name}>` : `[<${name}>]`))
    .join(" ");
}

export function createUsageString(name: string, command: CommandConfig) {
  const positionals = createPositionalUsage(command);
  return (
    `Usage: ${name} ${command.name}` +
    (positionals ? ` ${positionals}` : "") +
    (command.namedOptions ? " [--options]" : "")
  );
}

export function createOptionHelpString(option: CommandOption) {
  return `  - ${option.name}${option.help ? ` - ${option.help}` : ""}`;
}

export function createSubCommandHelp(name: string, command: CommandConfig) {
  let lines = [
    chalk`{bold ${command.name}} - ${command.help} `,
    "\n",
    createUsageString(name, command),
    command.manual ? "\n" + wrap(command.manual, { width: 80 }) : ""
  ];
  if (command.positionalOptions) {
    lines = [...lines, "", chalk.bold("Arguments:")];
    lines = [
      ...lines,
      ...(command.positionalOptions || []).map(createOptionHelpString)
    ];
  }
  if (command.namedOptions) {
    lines = [...lines, "", chalk.bold("Options:")];
    lines = [
      ...lines,
      ...(command.namedOptions || []).map(createOptionHelpString)
    ];
  }
  return lines.join("\n");
}

function findSubCommand(config: Config, name: string): ?CommandConfig {
  if (config.commands) {
    // $FlowFixMe
    return config.commands[name];
  } else {
    try {
      // $FlowFixMe
      return require(path.join(config.commandsPath, name));
    } catch (error) {
      return;
    }
  }
}

export async function help(config: Config, options: Options) {
  const dashes = "--------------------------";
  const subCommand = findSubCommand(config, options._[0]);
  logger.error(chalk`{gray ${dashes}} {bold ${config.name}} {gray ${dashes}}`);
  logger.error("");
  if (subCommand) {
    logger.error(createSubCommandHelp(config.name, subCommand));
  } else {
    logger.error(chalk.bold`Commands: `);
    logger.error(await createCommandsList(config));
  }
}
