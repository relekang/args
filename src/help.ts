/* eslint-disable global-require */
import chalk from 'chalk';
import wrap from 'word-wrap';
import path from 'path';
import fs from 'fs';
import { promisify } from 'util';

import { Config, CommandConfig, CommandOption, Options } from './types';

import * as logger from './logger';
import { findSubCommand } from './subCommand';

const readdir = promisify(fs.readdir);

function values<A>(o: { [key: string]: A }): Array<A> {
  return Object.values(o);
}

function findCommandDependencies(
  prefixes: string[],
  dependencies: { [key: string]: string } | undefined
) {
  if (!dependencies) {
    return {};
  }
  return Object.keys(dependencies)
    .filter(name => prefixes.find(prefix => name.indexOf(prefix) === 0))
    .reduce((lastValue, name) => {
      const command = require(name);
      return { ...lastValue, [command.name]: command };
    }, {});
}

async function getCommands(
  config: Config
): Promise<{ [key: string]: CommandConfig }> {
  let commands: { [key: string]: CommandConfig };
  // @ts-ignore
  if (config.commands) {
    // @ts-ignore
    commands = config.commands;
  } else {
    // @ts-ignore
    const commandsPath: string = config.commandsPath;
    commands = (await readdir(commandsPath))
      .filter(file => /\.[jt]s$/.test(file))
      .reduce(
        (lastValue, file) => ({
          ...lastValue,
          [file.replace(/.[jt]s$/, '')]: require(path.join(commandsPath, file)),
        }),
        {}
      );
  }
  if (
    config.commandPackagePrefixes &&
    config.commandPackagePrefixes.length > 0 &&
    config.packageInfo
  ) {
    commands = {
      ...commands,
      ...findCommandDependencies(
        config.commandPackagePrefixes,
        config.packageInfo.devDependencies
      ),
      ...findCommandDependencies(
        config.commandPackagePrefixes,
        config.packageInfo.dependencies
      ),
    };
  }
  return commands;
}

export async function createCommandsList(config: Config) {
  const commands = await getCommands(config);

  const length = Math.max(
    0,
    ...values(commands).map(({ name }) => name.length)
  );
  return values(commands)
    .map(command => {
      const spacing = new Array(Math.max(1, length - command.name.length + 1))
        .fill(' ')
        .join('');
      return `${command.name}${spacing} - ${command.help || ''}`;
    })
    .join('\n');
}

export function createPositionalUsage(command: CommandConfig) {
  return (command.positionalOptions || [])
    .map(({ name, required }) => (required ? `<${name}>` : `[<${name}>]`))
    .join(' ');
}

export function createUsageString(name: string, command: CommandConfig) {
  const positionals = createPositionalUsage(command);
  return (
    `Usage: ${name} ${command.name}` +
    (positionals ? ` ${positionals}` : '') +
    (command.namedOptions ? ' [--options]' : '')
  );
}

export function createOptionHelpString(option: CommandOption) {
  return `  - ${option.name}${option.help ? ` - ${option.help}` : ''}`;
}

export function createSubCommandHelp(name: string, command: CommandConfig) {
  let lines = [
    chalk`{bold ${command.name}} - ${command.help} `,
    '\n',
    createUsageString(name, command),
    command.manual ? '\n' + wrap(command.manual, { width: 80 }) : '',
  ];
  if (command.positionalOptions) {
    lines = [...lines, '', chalk.bold('Arguments:')];
    lines = [
      ...lines,
      ...(command.positionalOptions || []).map(createOptionHelpString),
    ];
  }
  if (command.namedOptions) {
    lines = [...lines, '', chalk.bold('Options:')];
    lines = [
      ...lines,
      ...(command.namedOptions || []).map(createOptionHelpString),
    ];
  }
  return lines.join('\n');
}

export async function help(config: Config, options: Options) {
  const dashes = '--------------------------';
  let subCommand;
  if (options._.length > 0) {
    subCommand = findSubCommand(config, options._[0]);
  }
  logger.error(chalk`{gray ${dashes}} {bold ${config.name}} {gray ${dashes}}`);
  logger.error('');
  if (subCommand) {
    logger.error(createSubCommandHelp(config.name, subCommand));
  } else {
    logger.error(chalk.bold`Commands: `);
    logger.error(await createCommandsList(config));
  }
}
