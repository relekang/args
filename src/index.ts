/* eslint-disable global-require */
import mri from 'mri';
import chalk from 'chalk';

import { help } from './help';
import * as logger from './logger';
import { CliError } from './errors';
import { parse } from './parse';
import { setup } from './setup';
import { Config, Options } from './types';
import { findSubCommand } from './subCommand';

export * from './errors';

async function __args(config: Config, subCommand: string, args: Array<string>) {
  await setup(config);
  if (subCommand === 'help') {
    return help(config, mri(args));
  }

  const command = findSubCommand(config, subCommand);

  if (command) {
    const options: Options = parse(command, args);
    await command.run(options);
  } else {
    throw new CliError({
      message: 'Unknown command',
      exitCode: 1,
      showHelp: true,
    });
  }
}

export function args(config: Config) {
  return ([_node, _program, subCommand, ...rest]: Array<string>) => {
    return __args(config, subCommand || config.defaultCommand || 'help', rest)
      .catch(async error => {
        if (
          error.constructor === CliError ||
          error.constructor.constructor === CliError
        ) {
          if (error.showHelp) {
            logger.error('\n');
            logger.error(chalk.bold(chalk.red(error.toString())));
            logger.error('\n\n');
            await help(config, mri(rest));
          } else {
            logger.error(error.toString());
          }
          if (process.env.NODE_ENV !== 'test') {
            process.exit(error.exitCode);
          }
        } else {
          logger.error(error);
        }
        if (process.env.NODE_ENV !== 'test') {
          process.exit(1);
        }
      })
      .catch(error => {
        logger.error(error.toString());
      });
  };
}

export default args;
