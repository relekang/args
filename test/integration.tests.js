// @flow
import args from "../src";
import * as logger from "../src/logger";
import type { Config } from "../src/types";

jest.mock("../src/logger", () => ({ error: jest.fn(), log: jest.fn() }));

const config: Config = {
  name: "supercli",
  commands: {
    test: {
      name: "test",
      help: "Testing testing",
      manual:
        "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse felis ligula, vulputate tincidunt consectetur sed, euismod sit amet dolor. Cras commodo eu mi sed consectetur. Quisque eget mauris felis. Sed accumsan quis dui quis consequat. Proin a magna mauris. Aenean sit amet mauris sem. Nam tellus eros, malesuada vel diam eget, consectetur sodales nisl. Quisque non libero auctor, tempor nisl eu, rutrum ante. Suspendisse eget lacus ex. ",
      run: jest.fn(),
      positionalOptions: [
        { name: "first", required: true },
        { name: "second", required: false }
      ],
      namedOptions: [{ name: "verbose", required: true }]
    },
    "long-named-command": {
      name: "long-named-command",
      help: "Command with long name",
      run: () => {}
    }
  }
};

beforeEach(() => {
  // $FlowFixMe flow does not understand mocks
  logger.error.mockReset();
  // $FlowFixMe flow does not understand mocks
  logger.log.mockReset();
});

test("args should parse the args", async () => {
  await args(config)(["node", "cli", "test"]);

  // $FlowFixMe
  expect(config.commands.test.run).toBeCalledWith({ _: [] });
});

test("args should show help screen", async () => {
  await args(config)(["node", "cli", "help"]);

  // $FlowFixMe flow does not understand mocks
  expect(logger.error.mock.calls.join("\n")).toMatchSnapshot();
});

test("args should show help screen for specific command", async () => {
  await args(config)(["node", "cli", "help", "test"]);

  // $FlowFixMe flow does not understand mocks
  expect(logger.error.mock.calls.join("\n")).toMatchSnapshot();
});
