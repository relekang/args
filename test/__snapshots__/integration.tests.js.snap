// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`args should show help screen 1`] = `
"[90m--------------------------[39m [1msupercli[22m [90m--------------------------[39m

[1mCommands: [22m
test                - Testing testing
long-named-command  - Command with long name"
`;

exports[`args should show help screen for specific command 1`] = `
"[90m--------------------------[39m [1msupercli[22m [90m--------------------------[39m

[1mtest[22m - Testing testing 


Usage: supercli test <first> [<second>] [--options]

  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse felis 
  ligula, vulputate tincidunt consectetur sed, euismod sit amet dolor. Cras 
  commodo eu mi sed consectetur. Quisque eget mauris felis. Sed accumsan quis dui 
  quis consequat. Proin a magna mauris. Aenean sit amet mauris sem. Nam tellus 
  eros, malesuada vel diam eget, consectetur sodales nisl. Quisque non libero 
  auctor, tempor nisl eu, rutrum ante. Suspendisse eget lacus ex. 

[1mArguments:[22m
  - first
  - second

[1mOptions:[22m
  - verbose"
`;
